sdk/src/main/java/io/opentelemetry/sdk/metrics/BaseInstrument.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.metrics;
+
+import io.opentelemetry.metrics.Instrument;
+import io.opentelemetry.metrics.LabelSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public abstract class BaseInstrument<B> implements Instrument<B> {
+
+  private final Map<LabelSet, B> boundCounters = new HashMap<>();
+
+  private final String name;
+  private final String description;
+  private final Map<String, String> constantLabels;
+  private final List<String> labelKeys;
+
+  protected BaseInstrument(
+      String name, String description, Map<String, String> constantLabels, List<String> labelKeys) {
+    this.name = name;
+    this.description = description;
+    this.constantLabels = constantLabels;
+    this.labelKeys = labelKeys;
+  }
+
+  @Override
+  public B bind(LabelSet labelSet) {
+    return getOrCreate(labelSet);
+  }
+
+  private B getOrCreate(LabelSet labelSet) {
+    B boundLongCounter = boundCounters.get(labelSet);
+    if (boundLongCounter == null) {
+      boundLongCounter = create(labelSet);
+      boundCounters.put(labelSet, boundLongCounter);
+    }
+    return boundLongCounter;
+  }
+
+  protected abstract B create(LabelSet labelSet);
+
+  @SuppressWarnings("rawtypes")
```

Comment:
Use constraints on the templated B argument to avoid this:
`public abstract class BaseInstrument<B extents BaseBoundInstrument > implements Instrument<B>`

Reply:
non-issue after removing the guts of this class.